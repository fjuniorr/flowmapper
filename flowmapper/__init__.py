"""flowmapper."""

from .flowmap import Flowmap

__all__ = (
    "__version__",
    "Flowmap"
    # Add functions and variables you want exposed in `flowmapper.` namespace here
)

__version__ = "0.0.0-9010"
