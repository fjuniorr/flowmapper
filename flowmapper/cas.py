from functools import cached_property

class CAS():
    """
    Class for CAS Registry Numbers® (CAS RN®) that accepts padded or non-padded strings
    """
    def __init__(self, cas: str):
        if not isinstance(cas, str):
            raise TypeError(f'cas should be a str, not {type(cas).__name__}')
        else:
            self.cas = cas.strip().lstrip('0')
            self.digits = tuple(int(d) for d in self.cas[0:-1].replace('-', ''))
    def __repr__(self):
        return self.cas
    def __eq__(self, other):
        if not self.cas and not other.cas:
                result = False
        else:
            result = self.cas == other.cas
        return result
    @cached_property
    def check_digit_expected(self):
        """
        Expected digit acording to https://www.cas.org/support/documentation/chemical-substances/checkdig algorithm
        """
        result = sum([
            index * value 
            for index, value in enumerate(self.digits[::-1], start=1)
        ]) % 10
        return result
    @property
    def check_digit_actual(self):
        int(self.cas[-1])
    @property
    def valid(self):
        """
        True if check if CAS number is valid acording to https://www.cas.org/support/documentation/chemical-substances/checkdig algorithm
        """
        return self.check_digit_actual == self.check_digit_expected
