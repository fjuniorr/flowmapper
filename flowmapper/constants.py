CONTEXT_MAPPING = {
        'Air/(unspecified)': 'air/unspecified',
        'Air/high. pop.': 'air/urban air close to ground',
        'Air/indoor': 'air/indoor',
        'Air/low. pop.': 'air/non-urban air or from high stacks',
        'Air/low. pop., long-term': 'air/low population density, long-term',
        'Air/stratosphere + troposphere': 'air/lower stratosphere + upper troposphere',
        'Emissions to air/': 'air/unspecified',
        'Emissions to air/high. pop.': 'air/urban air close to ground',
        'Emissions to air/low. pop.': 'air/non-urban air or from high stacks',
        'Emissions to air/low. pop., long-term': 'air/low population density, long-term',
        'Emissions to air/stratosphere + troposphere': 'air/lower stratosphere + upper troposphere',
        'Soil/(unspecified)': 'soil/unspecified',
        'Soil/agricultural': 'soil/agricultural',
        'Soil/industrial': 'soil/industrial',
        'Emissions to soil/': 'soil/unspecified',
        'Emissions to soil/agricultural': 'soil/agricultural',
        'Emissions to soil/forestry': 'soil/forestry',
        'Emissions to soil/industrial': 'soil/industrial',
        'Water/(unspecified)': 'water/unspecified',
        'Water/groundwater, long-term': 'water/ground-, long-term',
        'Water/ocean': 'water/ocean',
        'Water/river': 'water/surface water',
        'Emissions to water/': 'water/unspecified',
        'Emissions to water/groundwater': 'water/ground-',
        'Emissions to water/groundwater, long-term': 'water/ground-, long-term',
        'Emissions to water/lake': 'water/surface water',
        'Emissions to water/ocean': 'water/ocean',
        'Emissions to water/river': 'water/surface water',
        'Emissions to water/river, long-term': 'water/surface water',
        'Resources/': 'natural resource/in ground',
        'Resources/biotic': 'natural resource/biotic',
        'Resources/in air': 'natural resource/in air',
        'Resources/in ground': 'natural resource/in ground',
        'Resources/in water': 'natural resource/in water',
        'Resources/land': 'natural resource/land',
        'Raw/(unspecified)': 'natural resource/in ground',
        'Raw/in water': 'natural resource/in water',
    }

RANDOM_NAME_DIFFERENCES_MAPPING = {
    '2,4-D, dimethylamine salt': 'Dimethylamine',
    'Acids, unspecified': 'Acidity, unspecified',
    'Argon-40/kg': 'Argon-40',
    'Arsenic': 'Arsenic V',
    'Benzo(b,j,k)fluoranthene': 'Benzo(b)fluoranthene',
    'BOD5 (Biological Oxygen Demand)': 'BOD5, Biological Oxygen Demand',
    'BOD5 (Biological Oxygen Demand), NO': 'BOD5, Biological Oxygen Demand',
    'Chlorotoluron': 'Chlortoluron',
    'Chromium III': 'Chromium',
    'Coal, 26.4 MJ per kg': 'Coal, hard, unspecified, in ground',
    'Coal, 29.3 MJ per kg': 'Coal, hard, unspecified, in ground',
    'COD (Chemical Oxygen Demand)': 'COD, Chemical Oxygen Demand',
    'COD (Chemical Oxygen Demand), NO': 'COD, Chemical Oxygen Demand',
    'Dioxin, 2,3,7,8 Tetrachlorodibenzo-p-': 'Dioxins, measured as 2,3,7,8-tetrachlorodibenzo-p-dioxin',
    'Discarded fish, demersal': 'Discarded fish, demersal, to ocean',
    'Discarded fish, pelagic': 'Discarded fish, pelagic, to ocean',
    'Flurochloridone': 'Fluorochloridone',
    'Kaolin ore': 'Kaolinite, 24% in crude ore, in ground',
    'Lead-210/kg': 'Lead-210',
    'Metaldehyde (tetramer)': 'Metaldehyde',
    'Metolachlor, (S)': 'Metolachlor',
    'Nitrate compounds': 'Nitrate',
    'p-Xylene': 'Xylene',
    'Phenols, unspecified': 'Phenol',
    'Radioactive species, unspecified': 'Radioactive species, other beta emitters',
    'Radium-226/kg': 'Radium-226',
    'Radium-228/kg': 'Radium-228',
    'Sand, gravel and stone, extracted for use': 'Gravel, in ground',
    'Tri-allate': 'Triallate',
    '1-Butene': 'Butene',
    '2-Butene': 'Butene',
    'BTEX (Benzene, Toluene, Ethylbenzene, and Xylene), unspecified ratio': 'Benzene',
    'Calcium carbonate': 'Calcium, in ground',
    'Calcium chloride': 'Calcium, in ground',
    'Ethane, 1,1,2-trifluoro-, HFC-143': 'Ethane, 1,1,1-trifluoro-, HFC-143a',
    'Magnesium chloride': 'Magnesium, in ground',
    'Phosphate, NO': 'Phosphate',
    'Sulfur monoxide': 'Sulfur oxides',
    'Formetanate': 'Formetanate hydrochloride',
    'Haloxyfop': 'haloxyfop-ethoxyethyl',
    'Particulates, > 10 um': 'Particulates, > 2.5 um, and < 10um',
    'Plutonium': 'Plutonium-238',
    'Ruthenium': 'Ruthenium-103',
    'Particulates, < 10 um': 'Particulates, > 2.5 um, and < 10um',
    'Particulates, < 10 um (mobile)': 'Particulates, > 2.5 um, and < 10um',
    'Particulates, < 10 um (stationary)': 'Particulates, > 2.5 um, and < 10um',
    'Particulates, > 10 um (process)': 'Particulates, > 2.5 um, and < 10um',
    'Particulates, SPM': 'Particulates, > 2.5 um, and < 10um',
    'Particulates, diesel soot': 'Particulates, > 2.5 um, and < 10um',
    'Particulates, unspecified': 'Particulates, > 2.5 um, and < 10um',
    'Water/m3': 'Water',
    'Alkane (unspecified)': 'Hydrocarbons, aliphatic, alkanes, unspecified',
    'Formetanate': 'Formetanate hydrochloride',
    'Haloxyfop': 'haloxyfop-ethoxyethyl',
    'Bauxite': 'Gangue, bauxite, in ground',
    'Copper ore': 'Copper, Cu 0.2%, in mixed ore, in ground',
    'Copper, Cu 0.38%, Au 9.7E-4%, Ag 9.7E-4%, Zn 0.63%, Pb 0.014%, in ore': 'Copper, Cu 0.38%, in mixed ore, in ground',
    'Copper, Cu 3.2E+0%, Pt 2.5E-4%, Pd 7.3E-4%, Rh 2.0E-5%, Ni 2.3E+0% in ore': 'Copper, Cu 0.38%, in mixed ore, in ground',
    'Copper, Cu 5.2E-2%, Pt 4.8E-4%, Pd 2.0E-4%, Rh 2.4E-5%, Ni 3.7E-2% in ore': 'Copper, 0.52% in sulfide, Cu 0.27% and Mo 8.2E-3% in crude ore, in ground',
    'Coal, 18 MJ per kg': 'Coal, hard, unspecified, in ground',
    'Coal, brown, 10 MJ per kg': 'Coal, brown, in ground',
    'Coal, brown, 8 MJ per kg': 'Coal, brown, in ground',
    'Coal, hard': 'Coal, hard, unspecified, in ground',
    'Crude oil': 'Oil, crude, in ground',
    'Energy, from biomass': 'Energy, gross calorific value, in biomass',
    'Energy, from coal': 'Energy, gross calorific value, in biomass',
}

MINOR_LAND_NAME_DIFFERENCES_MAPPING = {
    'Transformation, to rivers, artificial': 'Transformation, to river, artificial',
    'Transformation, to lakes, artificial': 'Transformation, to lake, artificial',
    'Transformation, to grassland/pasture/meadow': 'Transformation, to grassland, natural, for livestock grazing',
    'Transformation, to agriculture, mosaic': 'Transformation, to heterogeneous, agricultural',
    'Transformation, from grassland/pasture/meadow': 'Transformation, from grassland, natural, for livestock grazing',
    'Transformation, from grassland': 'Transformation, from grassland, natural, for livestock grazing',
    'Occupation, lakes, artificial': 'Occupation, lake, artificial',
    'Occupation, grassland/pasture/meadow': 'Occupation, grassland, natural, for livestock grazing',
    'Occupation, agriculture': 'Occupation, annual crop',
    'Occupation, dump site, benthos': 'Occupation, seabed, unspecified',
    'Occupation, forest, intensive, normal': 'Occupation, forest, intensive',
    'Occupation, forest, intensive, short-cycle': 'Occupation, forest, intensive',
    'Occupation, forest, used': 'Occupation, forest, intensive',
    'Occupation, grassland/pasture/meadow': 'Occupation, pasture, man made',
    'Occupation, industrial area, benthos': 'Occupation, seabed, unspecified',
    'Occupation, industrial area, built up': 'Occupation, industrial area',
    'Occupation, industrial area, vegetation': 'Occupation, industrial area',
    'Occupation, permanent crop, fruit, intensive': 'Occupation, permanent crop, irrigated',
    'Occupation, sea and ocean': 'Occupation, seabed, unspecified',
    'Occupation, sea and ocean': 'Occupation, seabed, unspecified',
    'Occupation, traffic area': 'Occupation, traffic area, road network',
    'Occupation, traffic area, road embankment': 'Occupation, traffic area, road network',
    'Occupation, unspecified, used': 'Occupation, unspecified',
    'Occupation, water bodies, artificial': 'Occupation, river, artificial',
    'Occupation, water courses, artificial': 'Occupation, river, artificial',
    'Occupation, wetland': 'Occupation, inland waterbody, unspecified',
    'Transformation, from agriculture': 'Transformation, from annual crop',
    'Transformation, from annual crop, non-irrigated, fallow': 'Transformation, from pasture, man made',
    'Transformation, from forest, intensive, clear-cutting': 'Transformation, from forest, intensive',
    'Transformation, from forest, used': 'Transformation, from forest, intensive',
    'Transformation, from grassland': 'Transformation, from grassland, natural (non-use)',
    'Transformation, from grassland/pasture/meadow': 'Transformation, from pasture, man made',
    'Transformation, from industrial area, benthos': 'Transformation, from seabed, unspecified',
    'Transformation, from industrial area, built up': 'Transformation, from industrial area',
    'Transformation, from industrial area, vegetation': 'Transformation, from industrial area',
    'Transformation, from permanent crop, fruit': 'Transformation, from permanent crop, irrigated',
    'Transformation, from tropical rain forest': 'Transformation, from forest, extensive',
    'Transformation, from unspecified, used': 'Transformation, from unspecified',
    'Transformation, to agriculture': 'Transformation, to annual crop',
    'Transformation, to annual crop, fallow': 'Transformation, to arable land, unspecified use',
    'Transformation, to annual crop, non-irrigated, fallow': 'Transformation, to annual crop, non-irrigated, extensive',
    'Transformation, to dump site, benthos': 'Transformation, to seabed, unspecified',
    'Transformation, to forest, intensive, clear-cutting': 'Transformation, to forest, intensive',
    'Transformation, to forest, intensive, normal': 'Transformation, to forest, intensive',
    'Transformation, to forest, intensive, short-cycle': 'Transformation, to forest, intensive',
    'Transformation, to forest, used': 'Transformation, to forest, intensive',
    'Transformation, to grassland/pasture/meadow': 'Transformation, to pasture, man made',
    'Transformation, to industrial area, benthos': 'Transformation, to seabed, unspecified',
    'Transformation, to industrial area, built up': 'Transformation, to industrial area',
    'Transformation, to industrial area, vegetation': 'Transformation, to industrial area',
    'Transformation, to permanent crop, fruit, intensive': 'Transformation, to permanent crop, irrigated',
    'Transformation, to sea and ocean': 'Transformation, to seabed, unspecified',
    'Transformation, to traffic area, road embankment': 'Transformation, to traffic area, road network',
    'Transformation, to unspecified, used': 'Transformation, to unspecified',
    'Transformation, to urban/industrial fallow': 'Transformation, to industrial area',
    'Transformation, to water bodies, artificial': 'Transformation, to river, artificial',
    'Transformation, to water courses, artificial': 'Transformation, to river, artificial',
}

MISSING_FOSSIL_AND_BIOGENIC_CARBON_MAPPING = {
    "Carbon dioxide": "Carbon dioxide, fossil",
    "Carbon dioxide, biogenic": "Carbon dioxide, non-fossil",
    "Carbon dioxide, fossil": "Carbon dioxide, fossil",
    "Carbon dioxide, land transformation": "Carbon dioxide, from soil or biomass stock",
    "Carbon dioxide, peat oxidation": "Carbon dioxide, from soil or biomass stock",
    "Methane": "Methane, fossil",
    "Methane, biogenic": "Methane, non-fossil",
    "Methane, fossil": "Methane, fossil",
    "Methane, land transformation": "Methane, from soil or biomass stock",
    "Methane, peat oxidation": "Methane, from soil or biomass stock",
    "Carbon monoxide": "Carbon monoxide, fossil",
    "Carbon monoxide, biogenic": "Carbon monoxide, non-fossil",
    "Carbon monoxide, fossil": "Carbon monoxide, fossil",
    "Carbon monoxide, land transformation": "Carbon monoxide, from soil or biomass stock",
}

NAME_DIFFERENCES_WITH_UNIT_CONVERSION_MAPPING = {
        'Ammonia, as N': 'Ammonia',
        'Gas, mine, off-gas, process, coal mining/kg': 'Gas, mine, off-gas, process, coal mining',
        'AOX (Adsorbable Organic Halogens)': 'AOX, Adsorbable Organic Halogen as Cl',
        'AOX, Adsorbable Organic Halogen': 'AOX, Adsorbable Organic Halogen as Cl',
    }
