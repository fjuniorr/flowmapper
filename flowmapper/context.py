from dataclasses import asdict, dataclass

import flowmapper.jsonpath as jp


@dataclass
class Context:
    full: str = None
    primary: str = None
    secondary: str = None

    @classmethod
    def from_dict(cls, d, fields):
        fields = [fields] if isinstance(fields, str) else fields
        context_mapping_length = len(fields)
        primary_context = jp.extract(fields[0], d)
        if context_mapping_length == 1:
            result = Context(primary_context, primary_context)
        elif context_mapping_length == 2:
            secondary_context = jp.extract(fields[1], d)
            secondary_context = secondary_context if isinstance(secondary_context, str) else '/'.join(secondary_context)
            result = Context(
                full = f"{primary_context}/{secondary_context}",
                primary = primary_context,
                secondary = secondary_context
            )
        return result

    def to_dict(self):
        return asdict(self)
