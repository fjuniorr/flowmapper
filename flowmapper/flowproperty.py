from dataclasses import asdict, dataclass, field
import unicodedata
import flowmapper.jsonpath as jp

@dataclass
class FlowProperty:
    value: str = None
    raw_value: str = None
    raw_object: dict = field(default_factory=lambda: {})

    def __post_init__(self):
        raw_value = self.value
        self.value = self.normalize(self.value)
        if raw_value != self.value:
            self.raw_value = raw_value

    @classmethod
    def from_dict(cls, d, spec):
        if spec:
            key = jp.root(spec)
            value = jp.extract(spec, d)
            result = FlowProperty(
                value = cls.normalize(value),
                raw_value = value,
                raw_object = {key: d.get(key)},
            )
        else:
            result = FlowProperty(None)
        return result

    def to_dict(self):
        return asdict(self)

    def __eq__(self, other):
        result = False
        if self and other:
            if isinstance(other, FlowProperty):
                result = self.value == other.value
            else:
                result = self.value == other
        return result

    def __bool__(self):
        return bool(self.value)

    def __repr__(self):
        return str(self.value)

    def __hash__(self):
        return hash(self.value)

    @staticmethod
    def normalize(x):
        if x is None:
            result = None
        elif isinstance(x, str):
            result = normalize_str(x)
        else:
            result = [normalize_str(s) for s in x]
        return result

def normalize_str(s):
    return unicodedata.normalize('NFC', s).strip().lower()


class Name(FlowProperty):
    pass

class Synonyms(FlowProperty):
    pass