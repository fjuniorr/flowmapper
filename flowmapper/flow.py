import flowmapper.jsonpath as jp

from .unit import Unit
from .cas import CAS
from .context import Context
from .utils import generate_flow_id, transform_flow
from .flowproperty import FlowProperty


class Flow:
    def __init__(
        self,
        original,
        fields=None,
        transformations=None,
    ):
        transformed = (
            transform_flow(original, transformations) if transformations else original
        )
        fields = fields if fields else {k:k for k in original}
        uuid_spec = fields.get("uuid")
        name_spec = fields.get("name")
        synonyms_spec = fields.get("synonyms")
        context_spec = fields.get("context")
        unit_spec = fields.get("unit")
        cas_spec = fields.get("cas")

        self.id = generate_flow_id(original)
        self.uuid = jp.extract(uuid_spec, transformed)
        self.uuid_raw_value = jp.extract(uuid_spec, original)
        self.uuid_raw_object = jp.extract(uuid_spec, original, "object")
        self.name = FlowProperty.from_dict(transformed, name_spec)
        self.name_raw_value = jp.extract(name_spec, original)
        self.name_raw_object = jp.extract(name_spec, original, "object")
        self.synonyms = FlowProperty.from_dict(transformed, synonyms_spec)
        self.synonyms_raw_value = jp.extract(synonyms_spec, original)
        self.synonyms_raw_object = jp.extract(synonyms_spec, original, "object")
        self.context = Context.from_dict(transformed, context_spec)
        self.context_raw_value = jp.extract(context_spec, original)
        self.context_raw_object = jp.extract(context_spec, original, "object")
        self.unit = Unit.from_dict(transformed, fields["unit"])
        self.unit_raw_value = jp.extract(unit_spec, original)
        self.unit_raw_object = jp.extract(unit_spec, original, "object")
        self.cas = CAS(jp.extract(cas_spec, transformed))
        self.cas_raw_value = jp.extract(cas_spec, original)
        self.cas_raw_object = jp.extract(cas_spec, original, "object")
        self.raw = original
        self.transformed = transformed

    def __repr__(self) -> str:
        return f"{self.name} (in {self.unit.value}) <{self.context.value}>"

    def __eq__(self, other):
        return self.id == other.id

    def __hash__(self):
        return hash(self.id)

    def __lt__(self, other):
        return self.name.raw_value < other.name.raw_value
