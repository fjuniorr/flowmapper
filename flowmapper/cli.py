import typer
from typing_extensions import Annotated
from pathlib import Path
import logging
from typing import Optional
from typing_extensions import Annotated
import importlib.metadata
import logging
import json
from pathlib import Path
from enum import Enum
from randonneur import migrate_datasets
from .flow import Flow
from .utils import read_field_mapping, read_flowlist, read_migration_files
from .flowmap import Flowmap
try:
    import tomllib
except ModuleNotFoundError:
    import tomli as tomllib

logger = logging.getLogger(__name__)

app = typer.Typer()

def version_callback(value: bool):
    if value:
        print(f"flowmapper, version {importlib.metadata.version('flowmapper')}")
        raise typer.Exit()

class OutputFormat(str, Enum):
    all = 'all'
    glad = 'glad'
    randonneur = 'randonneur'

@app.callback()
def main(version: Annotated[
        Optional[bool],
        typer.Option("--version", callback=version_callback, is_eager=True),
    ] = None,):
    """
    Generate mappings between elementary flows lists
    """

@app.command()
def map(
    source: Annotated[Path, typer.Argument(help='Path to source flowlist')],
    target: Annotated[Path, typer.Argument(help='Path to target flowlist')],
    fields: Annotated[Path, typer.Option(help='Relationship between fields in source and target flowlists')],
    output_dir: Annotated[Path, typer.Option(help='Directory to save mapping and diagnostics files')] = Path('.'),
    format: Annotated[OutputFormat, typer.Option(help='Mapping file output format', case_sensitive=False)] = 'all',
    transformations: Annotated[list[Path], typer.Option("--transformations", "-t", help='Randonneur data migration file with changes to be applied to source flows before matching')] = None,
    unmatched_source: Annotated[bool, typer.Option(help='Write original source unmatched flows into separate file?')] = True,
    unmatched_target: Annotated[bool, typer.Option(help='Write original target unmatched flows into separate file?')] = True,
    matched_source: Annotated[bool, typer.Option(help='Write original source matched flows into separate file?')] = False,
    matched_target: Annotated[bool, typer.Option(help='Write original target matched flows into separate file?')] = False,
):
    """
    Generate mappings between elementary flows lists
    """
    output_dir.mkdir(parents=True, exist_ok=True)
    field_mapping = read_field_mapping(fields)
    if transformations:
            transformations = read_migration_files(*transformations)
    source_flows = [Flow(flow, field_mapping['source'], transformations) for flow in read_flowlist(source)]
    target_flows = [Flow(flow, field_mapping['target']) for flow in read_flowlist(target)]

    flowmap = Flowmap(source_flows, target_flows)
    flowmap.statistics()

    stem = f'{source.stem}-{target.stem}'

    if matched_source:
        with open(output_dir / f'{stem}-matched-source.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.matched_source], fs, indent=True)

    if unmatched_source:
        with open(output_dir / f'{stem}-unmatched-source.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.unmatched_source], fs, indent=True)

    if matched_target:
        with open(output_dir / f'{stem}-matched-target.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.matched_target], fs, indent=True)

    if unmatched_target:
        with open(output_dir / f'{stem}-unmatched-target.json', 'w') as fs:
            json.dump([flow.raw for flow in flowmap.unmatched_target], fs, indent=True)

    if format.value == 'randonneur':
        with open(output_dir / f'{stem}.json', 'w') as fs:
            json.dump(flowmap.to_randonneur(), fs, indent=2)
    elif format.value == 'glad':
        flowmap.to_glad().to_excel(output_dir / f'{stem}.xlsx', index = False)
    else:
        with open(output_dir / f'{stem}.json', 'w') as fs:
            json.dump(flowmap.to_randonneur(), fs, indent=2)
        flowmap.to_glad().to_excel(output_dir / f'{stem}.xlsx', index = False)
