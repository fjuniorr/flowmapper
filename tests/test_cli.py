from typer.testing import CliRunner
from flowmapper.cli import app
import json
import pandas as pd

runner = CliRunner()


def test_version():
    result = runner.invoke(app, ["--version"])
    assert result.output.startswith("flowmapper, version")


def test_format_glad(tmp_path):
    result = runner.invoke(
        app,
        [
            "map",
            "tests/data/sp.json",
            "tests/data/ei-3.7.json",
            "--format",
            "glad",
            "--fields",
            "tests/data/field_mapping-sp-ei.py",
            "--output-dir",
            str(tmp_path),
        ],
    )
    expected_files = sorted(
        [
            tmp_path / "sp-ei-3.7.xlsx",
            tmp_path / "sp-ei-3.7-unmatched-source.json",
            tmp_path / "sp-ei-3.7-unmatched-target.json",
        ]
    )

    files = sorted(tmp_path.glob("**/*"))

    assert result.exit_code == 0
    assert expected_files == files


def test_format_randonneur(tmp_path):
    result = runner.invoke(
        app,
        [
            "map",
            "tests/data/sp.json",
            "tests/data/ei-3.7.json",
            "--format",
            "randonneur",
            "--fields",
            "tests/data/field_mapping-sp-ei.py",
            "--output-dir",
            str(tmp_path),
        ],
    )
    expected_files = sorted(
        [
            tmp_path / "sp-ei-3.7.json",
            tmp_path / "sp-ei-3.7-unmatched-source.json",
            tmp_path / "sp-ei-3.7-unmatched-target.json",
        ]
    )

    files = sorted(tmp_path.glob("**/*"))

    assert result.exit_code == 0
    assert expected_files == files


def test_matched_flows(tmp_path, snapshot):
    runner.invoke(
        app,
        [
            "map",
            "tests/data/sp.json",
            "tests/data/ei-3.7.json",
            "--matched-source",
            "--matched-target",
            "--fields",
            "tests/data/field_mapping-sp-ei.py",
            "--output-dir",
            str(tmp_path),
        ],
    )

    with open(tmp_path / "sp-ei-3.7-matched-source.json") as fs:
        actual = json.load(fs)

    assert actual == snapshot


def test_matched_flows_with_randonneur_transformations(tmp_path, snapshot):
    runner.invoke(
        app,
        [
            "map",
            "tests/data/sp.json",
            "tests/data/ei-3.7.json",
            "--transformations",
            "tests/data/transformations.json",
            "--matched-source",
            "--matched-target",
            "--fields",
            "tests/data/field_mapping-sp-ei.py",
            "--output-dir",
            str(tmp_path),
        ],
    )

    with open(tmp_path / "sp-ei-3.7-matched-source.json") as fs:
        actual = json.load(fs)

    assert actual == snapshot


def test_matched_flows_with_multiple_randonneur_transformations(tmp_path, snapshot):
    runner.invoke(
        app,
        [
            "map",
            "tests/data/sp.json",
            "tests/data/ei-3.7.json",
            "--transformations",
            "tests/data/transformations.json",
            "--transformations",
            "tests/data/migrations.json",
            "--matched-source",
            "--matched-target",
            "--fields",
            "tests/data/field_mapping-sp-ei.py",
            "--output-dir",
            str(tmp_path),
        ],
    )

    with open(tmp_path / "sp-ei-3.7-matched-source.json") as fs:
        actual = json.load(fs)

    assert actual == snapshot
