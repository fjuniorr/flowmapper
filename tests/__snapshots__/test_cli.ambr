# serializer version: 1
# name: test_matched_flows
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        '(unspecified)',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, FR',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_matched_flows_with_multiple_randonneur_transformations
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        '(unspecified)',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        'high. pop.',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, FR',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, as N',
      'unit': 'kg',
    }),
  ])
# ---
# name: test_matched_flows_with_randonneur_transformations
  list([
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        '(unspecified)',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'CAS': '000110-63-4',
      'categories': list([
        'Air',
        'high. pop.',
      ]),
      'name': '1,4-Butanediol',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, FR',
      'unit': 'kg',
    }),
    dict({
      'categories': list([
        'Emissions to air',
        'low. pop.',
      ]),
      'name': 'Ammonia, as N',
      'unit': 'kg',
    }),
  ])
# ---
